### ISSUE TITLE
Qwen3.8 reasoning with MTP5 enters a repeated-token loop and never emits final content

### ISSUE THREAD

**Issue by mozdyn** [Neroued/ninfer#92]
# Bug: Qwen3.8 reasoning with MTP5 enters a repeated-token loop and never emits final content

## Summary

On the V100 build, Qwen3.8-27B NVFP4 with MTP5 works correctly in non-thinking mode, but reasoning mode can enter a deterministic repeated-token loop. The request consumes the full 8,192-token output budget, returns empty `content`, and ends with `finish_reason=length`.

This reproduced with:

- MTP5 + full proposal head, explicit `reasoning_effort: "medium"`;
- MTP5 + full proposal head, artifact default reasoning effort;
- MTP5 + optimized proposal head (`--lm-head-draft`), artifact default reasoning effort.

The repository's published reasoning profile, MTP3 + `--lm-head-draft`, succeeds on the same prompt in 3.29 s. The same prompt also completes correctly in 0.85 s with `reasoning_effort: "none"`, MTP5 full-head acceptance 100%, and valid final JSON.

## Environment

- Repository: `geoffwatts/ninfer-v100`
- Commit: `bceed2921210c75ee438cf4d082b9d7e06ed82d5`
- GPU: Tesla V100 PCIe 32 GB (`sm_70`), max graphics clock 1380 MHz
- Driver: 580.173.02
- CUDA build/runtime: 12.8
- Artifact: `qwen3_8_27b_nvfp4.ninfer`
- Artifact SHA-256: `bb3360522a06e136e0367f5703414d26272b7285c8a6ab6194135c17dbd81b32`
- KV: INT8, automatic capacity
- Prefix reuse: disabled
- CUDA Graphs: enabled
- MTP: full proposal head, 5 draft tokens

## Server command

```bash
build-v100/apps/ninfer-serve qwen3_8_27b_nvfp4.ninfer \
  --host 127.0.0.1 \
  --port 18080 \
  --device 0 \
  --max-context 32768 \
  --kv-capacity auto \
  --kv-dtype int8 \
  --prefill-chunk 2048 \
  --max-concurrency 1 \
  --spec mtp \
  --draft-tokens 5 \
  --no-prefix-reuse \
  --request-log-jsonl request-log.jsonl
```

## Minimal reproduction: default reasoning

```bash
curl http://127.0.0.1:18080/v1/chat/completions \
  -H 'Content-Type: application/json' \
  -d '{
    "model": "qwen3.8-27b",
    "messages": [
      {
        "role": "system",
        "content": "You are being evaluated in a clean session. Follow the user request exactly."
      },
      {
        "role": "user",
        "content": "Zwróć wyłącznie poprawny JSON bez Markdownu i dodatkowych kluczy: status=\"degraded\", pliki=3, rollback=true, owner=null."
      }
    ],
    "temperature": 1.0,
    "top_p": 0.95,
    "top_k": 20,
    "min_p": 0.0,
    "presence_penalty": 0.0,
    "max_tokens": 8192,
    "seed": 240726,
    "stream": false
  }'
```

Explicit medium reproduces as well by adding:

```json
"reasoning_effort": "medium"
```

## Actual behavior

Default reasoning:

```text
elapsed: 637.96 s
finish_reason: length
prompt_tokens: 102
completion_tokens: 8192
content: ""
reasoning_content prefix:
  Użytkownik prosi o zwrócenie wyłącznie poprawnego JSON bez Markdownu i dodatkowych kluczy444444444444...
```

Server telemetry:

```text
finish=output_limit prompt=102 gen=8192
prefill=255.7 tok/s
decode=12.9 tok/s
wall=637.95 s
speculative=mtp 1.00 tok/round (0.0%)
```

Explicit `medium` produced the same failure on three separate PL prompts:

```text
636.16 s, 8192 tokens, empty final, finish=length
638.25 s, 8192 tokens, empty final, finish=length
638.58 s, 8192 tokens, empty final, finish=length
```

The repeated token was `4` in these reproductions.

## Control 1: published reasoning profile succeeds

Using the same artifact and prompt with the repository's documented reasoning profile:

```text
MTP3
--lm-head-draft
INT8 KV
prefill chunk 1024
default reasoning effort
```

Result:

```text
elapsed: 3.29 s
finish: stop token
completion tokens: 151
content: valid requested JSON
reasoning_content: finite and coherent
decode: 51.7 tok/s
MTP: 2.80 tokens/round, 59.9% acceptance
```

## Control 2: non-thinking succeeds

Use the MTP5 full-head server and prompt, but send the registered non-thinking sampler and disable reasoning:

```json
{
  "reasoning_effort": "none",
  "temperature": 0.7,
  "top_p": 0.8,
  "top_k": 20,
  "min_p": 0.0,
  "presence_penalty": 1.5,
  "max_tokens": 512,
  "seed": 240726
}
```

Result:

```text
elapsed: 0.85 s
finish_reason: stop
completion_tokens: 35
reasoning_content: absent
content:
{
  "status": "degraded",
  "pliki": 3,
  "rollback": true,
  "owner": null
}

decode: 76.2 tok/s
MTP: 6.00 tokens/round, 100.0% acceptance
```

## Expected behavior

Reasoning mode should either:

1. produce finite reasoning followed by the requested final JSON; or
2. fail explicitly if the selected reasoning/MTP combination is unsupported.

It should not enter a repeated-token loop, consume the full output budget, and return empty final content.

## Scope established so far

Confirmed:

- failure with Qwen3.8 NVFP4 + reasoning + MTP5 using both full and optimized proposal heads;
- failure with both default effort and explicit `medium`;
- success with MTP3 + optimized proposal head under default reasoning;
- success with MTP5 full head when reasoning is disabled;
- normal non-thinking MTP5 performance and up to 98–100% acceptance on matched workloads.

Not yet isolated:

- reasoning with MTP disabled;
- reasoning with MTP1, MTP2, or MTP4;
- groupwise-int artifact;
- whether the cause is an MTP5 state/graph boundary, sampling, or another width-specific interaction.

## Suggested next isolation checks

- run the same request with MTP disabled;
- compare MTP1–5 under the same seed;
- dump rendered prompt/token IDs around the transition into the repeated token;
- compare target-only logits against MTP verification logits at the first divergent step;
- verify GDN/MTP state publication and CUDA Graph replay for reasoning-mode prompts.

**Comment by mozdyn** [comment]
Sorry, it's fork related.
